src/SAML2/XML/saml/AuthnContextDecl.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SAML2\XML\saml;
+
+use DOMElement;
+use SAML2\Constants;
+use SAML2\DOMDocumentFactory;
+use SAML2\Utils;
+use SAML2\XML\Chunk;
+use Webmozart\Assert\Assert;
+
+/**
+ * Class representing SAML2 AuthnContextDecl
+ *
+ * @author Tim van Dijen, <tvdijen@gmail.com>
+ * @package SimpleSAMLphp
+ */
+class AuthnContextDecl extends \SAML2\XML\AbstractConvertable
+{
+    /** @var \SAML2\XML\Chunk */
+    private $decl;
+
+
+    /**
+     * Initialize an AuthnContextDecl.
+     *
+     * @param \SAML2\XML\Chunk $decl
+     */
+    public function __construct(Chunk $decl)
+    {
+        $this->setDecl($decl);
+    }
+
+
+    /**
+     * Collect the value of the decl-property
+     *
+     * @return \SAML2\XML\Chunk
+     *
+     * @throws \InvalidArgumentException if assertions are false
+     */
+    public function getDecl(): Chunk
+    {
+        return $this->decl;
+    }
+
+
+    /**
+     * Set the value of the decl-property
+     *
+     * @param \SAML2\XML\Chunk $decl
+     * @return void
+     */
+    private function setDecl(Chunk $decl): void
+    {
+        $this->decl = $decl;
+    }
+
+
+    /**
+     * Convert XML into a AuthnContextDecl
+     *
+     * @param \DOMElement $xml The XML element we should load
+     * @return \SAML2\XML\saml\AuthnContextDecl
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        Assert::same($xml->tagName, 'saml:AuthnContextDecl');
+        Assert::same($xml->namespaceURI, Constants::NS_SAML);
+
+        /** @psalm-var \DOMElement $xml->childNodes[1] */
+        return new self(new Chunk($xml->childNodes[1]));
```

Comment:
I was already thinking about a problem with this before seeing your implementation, and now I see it confirmed 😞. The contents of this element can be anything, meaning these two would be valid:

```xml
<saml:AuthnContextDecl>
  <foo>bar</foo>
</saml:AuthnContextDecl>
```

```xml
<saml:AuthnContextDecl>
  <foo>1</foo>
  <bar>2</bar>
</saml:AuthnContextDecl>
```

Both are valid, but the current implementation would only work with the former. The latter would yield an object that contains only the `<foo>` element. Even worse, this would also be valid:

```xml
<saml:AuthnContextDecl>
  foo
  <bar>2</bar>
</saml:AuthnContextDecl>
```

I don't see an easy way out, but I draw some conclusions:

1. We can't assume the number of nodes children of the element to be only 1.
2. We can't use `SAML2\XML\Chunk` with the contents, because that's intended for XML elements, but the contents can be of any type, and in an arbitrary number.

I think probably the best solution would be to recursively import all the child nodes into a new `DOMDocument` and append them there. The `DOMDocument` would then be the property we save, and the class needs to implement `\Serializable` the same way `SAML2\XML\Chunk` does it.

Reply:
Now I see why nobody ever implemented AuthnContextDecl and StatusDetail :rofl:

Reply:
😆

Reply:
https://stackoverflow.com/questions/5389076/difference-similarities-between-xsdany-and-xsdanytype
I think the 3rd case is out of the question, so that would leave us with an array of Chunks here as well

Reply:
Should be done.. Please take a look